File: recieve.py
import socket
import struct
import hashlib
import sys

PAYLOAD_LENGTH = 1024
PACKET_SIZE = PAYLOAD_LENGTH + 16 + 32


def receive_file(port):
    sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    sock.bind(("", port))

    received_data = {}
    sequence_number = 0
    max_sequence_number = float("inf")
    packet_count = 0
    TRANSMISSION_ID = None
    file_name = ""

    print(f"Listening on port {port}...")

    while True:
        packet, addr = sock.recvfrom(PACKET_SIZE)
        transmission_id, sequence_number = struct.unpack("!HI", packet[:6])

        if transmission_id is None:
            TRANSMISSION_ID = transmission_id
        elif transmission_id != TRANSMISSION_ID:
            continue

        if sequence_number == 0:
            max_sequence_number = struct.unpack("!I", packet[6:10])[0]
            file_name = packet[10:].decode("utf-8")
            sock.sendto(struct.pack("!HI", transmission_id, sequence_number), addr)
        elif sequence_number == max_sequence_number:
            file_md5 = packet[6:]
            sock.sendto(struct.pack("!HI", transmission_id, sequence_number), addr)
            break
        else:
            if sequence_number not in received_data:
                received_data[sequence_number] = packet[6:]
                packet_count += 1
            sock.sendto(struct.pack("!HI", transmission_id, sequence_number), addr)
            print(f"sequence_number: {sequence_number}")

    # Reconstruct the file
    if file_name == "":
        print("Error reconstructing file name")
        exit()
    with open(file_name, "wb") as f:
        for i in sorted(received_data.keys()):
            f.write(received_data[i])

    print("File has been reconstructed. Verifying integrity...")
    verify_file_integrity(file_name, file_md5)
    sock.close()


def verify_file_integrity(file_path, original_md5):
    md5 = hashlib.md5()
    with open(file_path, "rb") as f:
        for chunk in iter(lambda: f.read(4096), b""):
            md5.update(chunk)

    if md5.digest() == original_md5:
        print("File integrity verified successfully.")
    else:
        print("File integrity verification failed.")
        print(md5.hexdigest())


if __name__ == "__main__":
    if len(sys.argv) != 2:
        print("Usage: python receiver.py <port>")
        sys.exit(1)

    # ip = sys.argv[1]
    port = int(sys.argv[1])
    receive_file(port)

File: recieve.go
package main

import (
	"bytes"
	"crypto/md5"
	"encoding/binary"
	"fmt"
	"io"
	"net"
	"os"
	"strconv"
	"time"
)

const (
	PAYLOAD_LENGTH = 1024
	PACKET_SIZE    = PAYLOAD_LENGTH + 6
	TIMEOUT        = 100 * time.Millisecond
)

func receiveFile(port int) {
	// addr, err := net.ResolveUDPAddr("udp", fmt.Sprintf(":%d", port))
	// if err != nil {
	// 	panic(err)
	// }
	addr := net.UDPAddr{
		IP:   net.ParseIP("0.0.0.0"),
		Port: port,
	}
	conn, err := net.ListenUDP("udp", &addr)
	if err != nil {
		fmt.Printf("Listen err %v", err)
	}
	defer conn.Close()

	buffer := make([]byte, PACKET_SIZE)
	var transmissionID uint16
	var sequenceNumber, maxSequenceNumber uint32
	var fileName string
	var TRANSMISSIONID uint16
	receivedData := make(map[uint32][]byte)

	fmt.Printf("Listening on port %d...\n", port)
	fmt.Println()

	conn.SetReadDeadline(time.Now().Add(10 * time.Second))
	for {
		n, raddr, err := conn.ReadFromUDP(buffer)
		if err != nil {
			fmt.Println("Timeout: file information not received")
			return
		}
		if sequenceNumber == 0 {
			TRANSMISSIONID = binary.BigEndian.Uint16(buffer[0:2])
			sequenceNumber = binary.BigEndian.Uint32(buffer[2:6])
			maxSequenceNumber = binary.BigEndian.Uint32(buffer[6:10])
			fileName = string(buffer[10:n])
			fmt.Printf("Receiving file: %s\n", fileName)
			fmt.Println()
			// fmt.Printf("sequenceNumber: %v\n", sequenceNumber)
			ackPacket := make([]byte, 6)
			binary.BigEndian.PutUint16(ackPacket[0:2], transmissionID)
			binary.BigEndian.PutUint32(ackPacket[2:6], sequenceNumber)
			conn.WriteToUDP(ackPacket, raddr)
			break
		}
	}

	var fileMd5 []byte
	conn.SetReadDeadline(time.Now().Add(TIMEOUT))
	for {
		n, raddr, err := conn.ReadFromUDP(buffer)
		if err != nil {
			if netErr, ok := err.(net.Error); ok && netErr.Timeout() {
				conn.SetReadDeadline(time.Now().Add(TIMEOUT))
				continue
			}
		}
		transmissionID = binary.BigEndian.Uint16(buffer[0:2])
		sequenceNumber = binary.BigEndian.Uint32(buffer[2:6])
		// fmt.Printf("sequenceNumber: %v\n", sequenceNumber)

		if transmissionID != TRANSMISSIONID {
			continue
		}

		if sequenceNumber == maxSequenceNumber {
			fileMd5 = buffer[6:n]
			ackPacket := make([]byte, 6)
			binary.BigEndian.PutUint16(ackPacket[0:2], transmissionID)
			binary.BigEndian.PutUint32(ackPacket[2:6], sequenceNumber)
			conn.WriteToUDP(ackPacket, raddr)
			break
		} else {
			receivedData[sequenceNumber] = make([]byte, n-6)
			copy(receivedData[sequenceNumber], buffer[6:n])
			ackPacket := make([]byte, 6)
			binary.BigEndian.PutUint16(ackPacket[0:2], transmissionID)
			binary.BigEndian.PutUint32(ackPacket[2:6], sequenceNumber)
			conn.WriteToUDP(ackPacket, raddr)
		}
	}

	file, err := os.Create(fileName)
	if err != nil {
		panic(err)
	}
	defer file.Close()

	for i := uint32(1); i < maxSequenceNumber; i++ {
		file.Write(receivedData[i])
	}

	println("File received, verifying integrity...")

	verifyFileIntegrity(fileName, fileMd5)
}

func verifyFileIntegrity(filePath string, originalMd5 []byte) {
	file, err := os.Open(filePath)
	if err != nil {
		panic(err)
	}
	defer file.Close()

	hash := md5.New()
	if _, err := io.Copy(hash, file); err != nil {
		panic(err)
	}
	md5 := hash.Sum(nil)

	if bytes.Equal(md5, originalMd5) {
		fmt.Println("File integrity verified successfully.")
	} else {
		fmt.Println("File integrity verification failed.")
	}
}

func main() {
	if len(os.Args) != 2 {
		fmt.Println("Usage: go run receive.go <port>")
		os.Exit(1)
	}

	port, err := strconv.Atoi(os.Args[1])
	if err != nil {
		fmt.Println("Invalid port number")
		os.Exit(1)
	}
	receiveFile(port)
}
